MIR-Directed-Research/intelligent-music-recommender
# Support entity addition via Knowledge Base API, automate test DB setup, and add schema constraints

### Overview
This PR delivers:
- [x] Automatic setup and teardown of testing database
  - Stopped tracking the `test.db` file, since it's no longer necessary
  - All tests and the cli dev app use this functionality
- [x] API functions for adding songs, artists, edges, and nodes
  - Addition of nodes is done indirectly and transactionally as part of the addition of songs and artists
- [x] Schema constraints + testing
  - e.g. foreign key constraints now enabled, `NOT NULL`, range checks, etc.

Resolves #5
Resolves #6 
Resolves #9 
Resolves #18

### Notes
#### Test Database Setup and Teardown
Ideally, we could reduce the overhead of setting up and tearing down the database after each test method; a couple ideas include:
* create and remove database per test class instead of once per test method
  * use `setUpClass` and `tearDownClass` instead of `setUp` and `tearDown`
* optionally clear table contents in `tearDown` instead of altogether removing the database

See issue #17 

However, I think it's fine for now. The overhead is small (tests run in about a second) and the result is pretty independence between all tests.

#### PR Best Practices
Counter to best practices, I'm conflating multiple deliverables into a single PR. Delivering (tested + documented) functionality is currently our highest priority, so I'll disregard best PR practices this time.
